vendor: update k8s dependencies and fix allocator bug

aanm opened the pull request:
**Summary of changes**:
Update kubernetes dependencies which has caused to add functionality around the IPv4 gateway assigned to `cilium_host` between restarts. Please review per-commit

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4128)
<!-- Reviewable:end -->

aanm committed:
vendor: update k8s dependencies to 1.10.2

* For the cri integration the following dependencies were prematurely updated:

- All kubernetes libraries were updated to 1.10.2

* All transient dependencies were automatically delt by dep.

* pkg/ipam: fixed imports due kubernetes library update

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
common: add C2GoArray function

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/ip: add GetNextIP

Signed-off-by: André Martins <andre@cilium.io>

raybejjani commented on pkg/ip/ip.go:
```diff
@@ -337,6 +337,12 @@ func getPreviousIP(ip net.IP) net.IP {
 
+// GetNextIP returns the next IP from the given IP address. If the given IP is
+// the last IP of a v4 or v6 range, the same IP is returned.
+func GetNextIP(ip net.IP) net.IP {
```
Is there a reason we couldn't just make the getNextIP below public?

aanm replied:
It's fine by me, I'll add it to the commit

raybejjani commented on pkg/node/node_address.go:
```diff
@@ -402,3 +415,82 @@ func GetNodeAddressing(enableIPv4 bool) *models.NodeAddressing {
 	}
 }
+
+func getCiliumHostIPsFromFile(nodeConfig string) (ipv4GW, ipv6Router net.IP) {
```
Could you add a comment mentioning that this comes from node_config.h and maybe where to find the format it's parsing? Where do we write it out anyway?

raybejjani commented on common/utils.go:
```diff
@@ -40,6 +41,31 @@ func goArray2C(array []byte) string {
+// Example:
+// str := "0x12, 0xff, 0x0, 0x1"
+// fmt.Print(C2GoArray(str)) //`{0x12, 0xFF, 0x0, 0x01}`"
+func C2GoArray(str string) []byte {
```
Could you rename this to something that includes the word hex? (hex2GoArray or parseHexValues or whatever). I was expecting this to use some cgo things to make a copy of an actual C array. Unless C means something else here?

raybejjani replied:
oh, I se the goArray2C above. That's a really unfortunate name.

raybejjani replied:
Can this use https://golang.org/pkg/encoding/hex/#DecodeString ? Instead of splitting on the `,` you would replace it (and space) with nothing, then pass it along.

aanm replied:
Does not work.

Hint: Is `0x10, 0x1` equal to `101` or `1001`? :-)

raybejjani replied:
sigh... also omg

raybejjani commented on common/utils.go:
```diff
@@ -40,6 +41,31 @@ func goArray2C(array []byte) string {
+		strDigit := strings.TrimPrefix(hexDigit, "0x")
+		digit, err := strconv.ParseInt(strDigit, 16, 9)
+		if err != nil {
+			fmt.Printf("err %+v\n", err)
```
Are these printf calls on purpose? or were they for debug?

aanm replied:
debug, clean up

raybejjani commented on pkg/node/node_address.go:
```diff
@@ -402,3 +415,82 @@ func GetNodeAddressing(enableIPv4 bool) *models.NodeAddressing {
+		return
+	}
+	ipv4GW, ipv6Router = getCiliumHostIPsFromNetDev(HostDevice)
+	return
```
We don't do these naked returns in our code too often. It might still be good to return the variables directly despite it being redundant (I do prefer named return variables, however!)

raybejjani reviewed (changes requested):
Most of my feedback is naming and formatting. Otherwise, though, the PR title is a little misleading. Unless I'm reading it wrong, this PR also changes startup behaviour to better recover the node's IP info on startup (by reading node_config.h). Is that correct or did I misunderstand the code?

Relatedly, the actual vendor change commit will compile, right? I'm worried that a bisect might be harder in the case it needs the other commits to work but that might be unavoidable.

The `pkg/ip: add GetNextIP` commit seems a little unneeded? it could just be part of the commit where it is used but if you need to change more files (if you make the private function public) then it might be big enough to be its own commit. shrug. I would prefer changing the private function and keeping the commit instead of merging the commit that adds the public wrapper.

raybejjani reviewed (approved):
Thank you!

aanm committed:
pkg: allocate first IP in IPv4 allocation range

Since kubernetes allocator fixed their randomize function to allocate a
different IP on each new allocation we have to make sure existing cilium
instances keep the previously IPv4 gateway IP.

On a new node however, the IPv4 gateway IP will be the first IP from the
allocation range.

Fixes: cfe05650d (vendor: update k8s dependencies to 1.10.2)

Signed-off-by: André Martins <andre@cilium.io>
